File: Dots_[Front_+_Back]/dots.js
function add_dots(x){
    let final = []
    if(x.length <= 1){
        return x
    }else{
        for(let i of add_dots(x.slice(1,))){
            final.push(x[0] + i)
            final.push(x[0] + "." + i)
        }
        return final
    }
}

console.log(add_dots('abcd'));

File: Dots_[Front_+_Back]/dots.py
x = input()

def add_dots(x):
    final = []
    if len(x) <= 1:
        return x
    else:
        for i in add_dots(x[1:]):
            final.append(x[0] + i)
            final.append(x[0] + "." + i)
        return final

print(sorted(add_dots(x), key=lambda x: len(x)))
